roofit/histfactory/src/HistFactoryNavigation.cxx

```diff
@@ -753,25 +736,20 @@ namespace RooStats {
       RooArgSet allTerms;
 
       // Get All Subnodes of this product
-      RooArgSet productComponents = node->components();
+      //RooArgSet productComponents = node->components();
 
       // Loop over the subnodes and add
-      TIterator* itr = productComponents.createIterator();
-      RooAbsArg* arg = NULL;
-      while( (arg=(RooAbsArg*)itr->Next()) ) {
-   std::string ClassName = arg->ClassName();
-   if( ClassName == "RooProduct" ) {
-     RooProduct* prod = dynamic_cast<RooProduct*>(arg);
-     allTerms.add( _GetAllProducts(prod) );
-   }
-   else {
-     allTerms.add(*arg);
-   }
+      for (auto *arg : dynamic_range_cast<RooAbsArg *>(node->components())) {
+        std::string ClassName = arg->ClassName();
+        if( ClassName == "RooProduct" ) {
+            RooProduct* prod = dynamic_cast<RooProduct*>(arg);
```

Comment:
```suggestion
            RooProduct* prod = static_cast<RooProduct*>(arg);
```
Can you change this to `static_cast`? The previous line already makes sure that this is a `RooProduct`.